README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# wellsolve

<!-- badges: start -->

[![Travis build
status](https://travis-ci.com/teldosas/wellsolve.svg?branch=master)](https://travis-ci.com/teldosas/wellsolve)
<!-- badges: end -->

The goal of wellsolve is to minimize the Total Pumping Cost from an
Aquifer to a Water Tank, Via a Pipe Network

## Installation

You can install the released version of wellsolve from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("wellsolve")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("teldosas/wellsolve")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(wellsolve)

projected <- "+proj=utm +zone=34"
x0 = 650000
y0 = 4400000

# the first value is the tank and the rest are wells
xc <- c() ; xc <- c(0,100,180,100,700,800,900) ; xc <- xc + x0 -10000
yc <- c() ; yc <- c(0,0,0,80,0,0,900) ; yc <- yc + y0 -10000

# xstart is a wild guess about the result.
# This helps the algorithm search for the result in the correct area
xstart<- c(0.250,0.1,0.25,0.1,0.1,0.01)

nw=7 #number of wells + 1 (the tank)
links<- data.frame(matrix(ncol=2, nrow=(nw-1)))
links[1:(nw-1),1]<-1:(nw-1)
links[1,2]<-0 ; links[2,2]<-1 ; links[3,2]<-0
links[4,2]<-2 ; links[5,2]<-4 ; links[6,2]<-2

Dlist<-list()
Dlist[1:6]<-0.05


# number of extra wells
nwextra <- 2
nwextra <- nwextra + nw
xextra<- c() ; xextra <- c(500,800) ; xextra <- xextra + x0 -10000
yextra<- c() ; yextra <- c(500,300) ; yextra <- yextra + y0 -10000
qextra<- c() ; qextra[1]<- 0; qextra[2]<- 0

result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)
```

    ##   Algorithm parameters
    ##   --------------------
    ##   Method: Broyden  Global strategy: double dogleg (initial trust region = -2)
    ##   Maximum stepsize = 1.79769e+308
    ##   Scaling: fixed
    ##   ftol = 1e-08 xtol = 1e-08 btol = 0.01 cndtol = 1e-12
    ## 
    ##   Iteration report
    ##   ----------------
    ##   Iter         Jac     Lambda      Eta     Dlt0     Dltn         Fnorm   Largest |f|
    ##      0                                                    5.462367e+11  7.410957e+05
    ##      1  N(6.8e-08) N            0.9898   0.1411   0.2822  3.263078e+10  1.769475e+05
    ##      2  B(8.8e-08) N            0.9893   0.0372   0.0372  8.693787e+09  7.796567e+04
    ##      3  B(5.3e-08) N            0.9960   0.0365   0.0730  1.061454e+09  3.210573e+04
    ##      4  B(3.8e-08) N            0.9321   0.0201   0.0201  3.871930e+08  1.946157e+04
    ##      5  B(2.3e-08) W   0.7108   0.9143   0.0201   0.0401  8.979170e+07  9.374749e+03
    ##      6  B(1.5e-08) N            0.9019   0.0318   0.0637  2.519811e+06  1.808672e+03
    ##      7  B(1.4e-08) N            0.9046   0.0030   0.0059  4.167393e+05  6.473717e+02
    ##      8  B(5.9e-09) N            0.9095   0.0049   0.0098  6.329708e+04  2.428495e+02
    ##      9  B(5.4e-09) N            0.9475   0.0004   0.0004  2.216752e+04  1.475947e+02
    ##     10  B(5.9e-09) N            0.8946   0.0001   0.0000  2.170522e+04  1.450197e+02
    ##     11  B(9.9e-10) W   0.0399   0.8497   0.0000   0.0000  2.108433e+04  1.411546e+02
    ##     12  B(5.6e-08) C            0.8873   0.0000   0.0000  2.049992e+04  1.395073e+02
    ##     13  B(7.1e-08) W   0.0021   0.2056   0.0000   0.0000  2.060405e+04  1.388003e+02
    ##     14  N(5.2e-07) N            0.2489   0.0017   0.0035  2.455097e+02  1.575401e+01
    ##     15  B(5.2e-07) N            0.2367   0.0002   0.0004  9.223712e+00  3.044597e+00
    ##     16  B(5.5e-07) N            0.6625   0.0001   0.0001  2.607577e-04  1.449910e-02
    ##     17  B(5.5e-07) N            0.3479   0.0000   0.0000  5.376011e-07  7.942852e-04
    ##     18  B(5.6e-07) N            0.3109   0.0000   0.0000  1.836902e-07  4.390565e-04
    ##     19  B(5.5e-07) N            0.2292   0.0000   0.0000  4.581097e-10  2.331423e-05

``` r
# Case A
Dlist[1:6]<-2
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)
```

    ##   Algorithm parameters
    ##   --------------------
    ##   Method: Broyden  Global strategy: double dogleg (initial trust region = -2)
    ##   Maximum stepsize = 1.79769e+308
    ##   Scaling: fixed
    ##   ftol = 1e-08 xtol = 1e-08 btol = 0.01 cndtol = 1e-12
    ## 
    ##   Iteration report
    ##   ----------------
    ##   Iter         Jac     Lambda      Eta     Dlt0     Dltn         Fnorm   Largest |f|
    ##      0                                                    2.191820e+05  3.933785e+02
    ##      1  N(2.0e-04) N            0.9605   0.2779   0.5558  2.460913e-05  3.316578e-03
    ##      2  B(2.0e-04) N            0.9948   0.0000   0.0000  3.059305e-14  1.164155e-07
    ##      3  B(2.0e-04) N            0.9990   0.0000   0.0000  6.342875e-26  1.928310e-13

``` r
# Case LC1
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)
```

    ##   Algorithm parameters
    ##   --------------------
    ##   Method: Broyden  Global strategy: double dogleg (initial trust region = -2)
    ##   Maximum stepsize = 1.79769e+308
    ##   Scaling: fixed
    ##   ftol = 1e-08 xtol = 1e-08 btol = 0.01 cndtol = 1e-12
    ## 
    ##   Iteration report
    ##   ----------------
    ##   Iter         Jac     Lambda      Eta     Dlt0     Dltn         Fnorm   Largest |f|
    ##      0                                                    2.191839e+05  3.933803e+02
    ##      1  N(2.0e-04) N            0.9605   0.2779   0.5558  5.736430e-13  8.167479e-07
    ##      2  B(2.0e-04) N            0.6520   0.0000   0.0000  9.087678e-28  2.842171e-14

``` r
# Case B
Dlist[1:6]<-0.4
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)
```

    ##   Algorithm parameters
    ##   --------------------
    ##   Method: Broyden  Global strategy: double dogleg (initial trust region = -2)
    ##   Maximum stepsize = 1.79769e+308
    ##   Scaling: fixed
    ##   ftol = 1e-08 xtol = 1e-08 btol = 0.01 cndtol = 1e-12
    ## 
    ##   Iteration report
    ##   ----------------
    ##   Iter         Jac     Lambda      Eta     Dlt0     Dltn         Fnorm   Largest |f|
    ##      0                                                    2.139104e+05  3.877135e+02
    ##      1  N(1.9e-04) N            0.9664   0.2662   0.5324  2.229651e+02  9.945352e+00
    ##      2  B(1.9e-04) N            0.9995   0.0068   0.0137  2.044975e+00  9.380462e-01
    ##      3  B(1.7e-04) N            0.9999   0.0006   0.0012  1.587382e-05  3.448635e-03
    ##      4  B(1.7e-04) N            0.9751   0.0000   0.0000  1.163745e-09  4.167468e-05
    ##      5  B(1.7e-04) N            0.6950   0.0000   0.0000  1.249868e-12  1.416031e-06
    ##      6  B(1.7e-04) N            0.6949   0.0000   0.0000  6.889993e-17  6.448097e-09

``` r
# Case C
qextra[1] <- 0.1
qextra[2] <- 0.4
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)
```

    ##   Algorithm parameters
    ##   --------------------
    ##   Method: Broyden  Global strategy: double dogleg (initial trust region = -2)
    ##   Maximum stepsize = 1.79769e+308
    ##   Scaling: fixed
    ##   ftol = 1e-08 xtol = 1e-08 btol = 0.01 cndtol = 1e-12
    ## 
    ##   Iteration report
    ##   ----------------
    ##   Iter         Jac     Lambda      Eta     Dlt0     Dltn         Fnorm   Largest |f|
    ##      0                                                    2.154684e+05  3.812010e+02
    ##      1  N(1.9e-04) N            0.9739   0.2619   0.5239  2.378423e+02  1.049493e+01
    ##      2  B(1.9e-04) N            0.9989   0.0071   0.0141  2.182377e+00  9.653594e-01
    ##      3  B(1.7e-04) N            0.9996   0.0006   0.0012  2.172868e-05  4.698379e-03
    ##      4  B(1.7e-04) N            0.7206   0.0000   0.0000  3.098375e-08  1.748498e-04
    ##      5  B(1.7e-04) N            0.6279   0.0000   0.0000  4.350881e-12  2.631109e-06
    ##      6  B(1.7e-04) N            0.5538   0.0000   0.0000  5.567208e-16  2.117073e-08

``` r
# T=0.0025 ;R=3000; ff=0.03; qtot<-0.5; r0=0.200
```


NEWS.md
# wellsolve 0.1.0

* First release on CRAN.


tests/testthat.R
library(testthat)
library(wellsolve)

test_check("wellsolve")


tests/testthat/test-wellsolve.R
# global state
projected <- "+proj=utm +zone=34"
x0 = 650000
y0 = 4400000

# the first value is the tank and the rest are wells
xc <- c() ; xc <- c(0,100,180,100,700,800,900) ; xc <- xc + x0 -10000
yc <- c() ; yc <- c(0,0,0,80,0,0,900) ; yc <- yc + y0 -10000

# xstart is a wild guess about the result.
# This helps the algorithm search for the result in the correct area
xstart<- c(0.250,0.1,0.25,0.1,0.1,0.01)

nw <- 7 #number of wells + 1 (the tank)
links<- data.frame(matrix(ncol=2, nrow=(nw-1)))
links[1:(nw-1),1]<-1:(nw-1)
links[1,2]<-0 ; links[2,2]<-1 ; links[3,2]<-0
links[4,2]<-2 ; links[5,2]<-4 ; links[6,2]<-2
links[7,1:2]<-c(5, 6)

Dlist<-list()
Dlist[1:6]<-0.05


# number of extra wells
nwextra <- 2
xextra<- c() ; xextra <- c(500,800) ; xextra <- xextra + x0 -10000
yextra<- c() ; yextra <- c(500,300) ; yextra <- yextra + y0 -10000
qextra<- c() ; qextra[1]<- 0; qextra[2]<- 0

test_that("wellsolve() works", {
  default_wellsolve <-
    function(ff) wellsolve(projected,
                           T=0.0025, R=3000, ff=ff, qtot<-0.5, r0=0.200,
                           Dlist, qextra, nw, xc, yc,links,
                           nwextra, xextra, yextra, xstart)

  result<-default_wellsolve(ff=0.03)

  # Case A
  Dlist[1:6]<-2
  result_a<-default_wellsolve(ff=0.03)

  # Case LC1
  result_lc1<-default_wellsolve(ff=0)

  # Case B
  Dlist[1:6]<-0.4
  result_b<-default_wellsolve(ff=0.03)

  # Case C
  qextra[1] <- 0.1
  qextra[2] <- 0.4
  result_c<-default_wellsolve(ff=0.03)

  expect_equal_to_reference(result, 'expected/result.rds')
  expect_equal_to_reference(result_a, 'expected/result_a.rds')
  expect_equal_to_reference(result_lc1, 'expected/result_lc1.rds')
  expect_equal_to_reference(result_b, 'expected/result_b.rds')
  expect_equal_to_reference(result_c, 'expected/result_c.rds')
})

test_that("calculate_distances() works", {
  xextra<- c(xextra,xc)
  yextra<- c(yextra,yc)

  result <- calculate_distances(xc, yc, projected, r0=0.2, nw=length(xc))
  resultextra <-
    calculate_distances(xextra, yextra, projected, r0=0.2, nw=length(xextra))
  expect_equal_to_reference(result$r, 'expected/r.rds')
  expect_equal_to_reference(result$coords, 'expected/coords.rds')
  expect_equal_to_reference(result$test, 'expected/test.rds')

  expect_equal_to_reference(resultextra$r, 'expected/rextra.rds')
  expect_equal_to_reference(resultextra$coords, 'expected/coordsextra.rds')
  expect_equal_to_reference(resultextra$test, 'expected/testextra.rds')
})

test_that("calculate_s() works", {
  # Case C
  qextra[1] <- 0.1
  qextra[2] <- 0.4

  qresult <- readRDS('mocks/qresult.rds')

  rnot <- readRDS('mocks/rnot.rds')
  rextranot <- readRDS('mocks/rextranot.rds')

  R <- 3000
  T <- 0.0025

  nwextra <- nwextra + nw
  s <- calculate_s(ncol=nw-1, nrow=nw-1, qresult, rnot, R, T)
  sextra <-
    calculate_s(ncol=nwextra-nw, nrow=nwextra, qextra, rextranot, R, T)

  expect_equal_to_reference(s, 'expected/s.rds')
  expect_equal_to_reference(sextra, 'expected/sextra.rds')
})


R/wellsolve.R
#' Minimize the Total Pumping Cost
#' @description  Minimize the Total Pumping Cost from an Aquifer
#' to a Water Tank, Via a Pipe Network
#'
#' @param projected A character string of projection arguments;
#' the arguments must be entered exactly as in the PROJ.4 documentation;
#' if the projection is unknown, use as.character(NA),
#' it may be missing or an empty string of zero length and
#' will then set to the missing value.
#' With rgdal built with PROJ >= 6 and GDAL >= 3,
#' the +init=key may only be used with value epsg:<code>.
#' @param T Aquifer’s transmissivity in square meters per second.
#' @param R Radius of influence of the system of wells in meters.
#' @param ff Friction coefficient along the pipes
#' @param qtot Total required flow rate in qubic meters per second.
#' @param r0 Diameter of all wells in meters.
#' @param Dlist A vector with the diameters of the pipes in meters.
#' @param qextra A vector with the required flow rates for the extra wells
#' in qubic meters per second.
#' @param nw Number of wells + 1 (the tank).
#' @param xc A vector with x coordinates. The first value is the tank and
#' the rest are the wells.
#' @param yc A vector with y coordinates. The first value is the tank and
#' the rest are the wells.
#' @param links A two-column matrix describing the links between the wells. Each
#' row describes one link. The two columns contain the id numbers of the
#' connected wells. The id number of the tank is 0.
#' @param nwextra Number of extra wells.
#' @param xextra A vector with x coordinates for the extra wells.
#' @param yextra A vector with y coordinates for the extra wells.
#' @param xstart A wild guess about the result.
#' This helps the algorithm search for the result in the correct area
#' @param g Acceleration of gravity.
#'
#' @return A list containing the following:
#' \item{qresult}{The optimal well flows}
#' \item{hf}{The hydraulic head losses}
#' \item{s}{The hydraulic head level drawdowns}
#' \item{printmap}{The diagram of the tank and the wells}
#' \item{sfinal}{The hydraulic head level drawdowns including
#' those of the extra wells}
#' \item{kx}{The xi constant of the Darcy-Weisbach formula}
#' \item{QQsqure}{The square of the pipe flow rates}
#' @export
#'
#' @example man/examples/example.R
#'
#' @references
#' Nagkoulis, N., Katsifarakis, K. Minimization of Total Pumping Cost from
#' an Aquifer to a Water Tank, Via a Pipe Network.
#' Water Resour Manage 34, 4147–4162 (2020).
#' \doi{10.1007/s11269-020-02661-x}
wellsolve <- function(projected, T, R,ff, qtot, r0, Dlist, qextra, nw,
                      xc, yc, links, nwextra, xextra, yextra, xstart,
                      g = 9.81) {
  temp <- calculate_distances(xc, yc, projected, r0, nw)
  r <- temp$r; coords <- temp$coords; test <- temp$test;

  nwextra <- nwextra + nw
  temp <-
    calculate_distances(c(xextra,xc), c(yextra,yc), projected, r0, nwextra)
  rextra <- temp$r; coordsextra <- temp$coords; testextra <- temp$test;

  lines <-list()
  names<-c()
  i=0
  while (i<nw-1) {
    i=i+1
    names<-c(names,paste0("line",i))
    var <- rbind(c(coords[i+1,1],coords[i+1,2]),c(coords[links[i,2]+1,1],coords[links[i,2]+1,2]))
    lines[[names[i]]] <-  raster::spLines(var, attr=data.frame(start=i,end=links[i,2]), crs=projected)
  }

  printmap <- render_diagram(test, lines, qextra, nw, testextra)

  Cmatrix <- data.frame(matrix(ncol=(nw-1), nrow=(nw-1)))
  Cmatrix[,] <-0

  i=0
  while (i<(nw-1)){
    i=i+1
    Cmatrix[[i,i]]<-1
    if (!(lines[[i]]@data[["end"]]==0)) {
      j <- lines[[i]]@data[["end"]]
      while (!(j==0)) {
        Cmatrix[[j,i]] <- 1
        j <- lines[[j]]@data[["end"]]
      }
    }
  }

  cn_ <- Cmatrix[,]-Cmatrix[,(nw-1)]
  ct <- t(cn_)

  kx <- data.frame(matrix(ncol=(nw-1), nrow=(nw-1)))
  kx[,] <-0
  i=0
  while (i<(nw-1)) {
    i=i+1
    lines[[i]]@data[[3]] <- r[lines[[i]]@data[[1]]+1,lines[[i]]@data[[2]]+1]
    # Darcy–Weisbach formula
    lines[[i]]@data[[4]] <-
      8*ff*lines[[i]]@data[[3]] / ( (Dlist[[i]]^5) * g *(pi)^2 )
    kx[[i,i]]<-lines[[i]]@data[[4]]
  }

  rnot<-r[2:nw,2:nw]
  a <- data.frame(matrix(ncol=(nw-1), nrow=(nw-1)))
  i=0
  while (i<(nw-1)) {
    i=i+1;  j=0
    while (j<(nw-1)) {
      j=j+1
      a[[i,j]] <- -log(rnot[[i,j]]/rnot[[j,(nw-1)]]) * (1/(2*pi*T))
    }
  }
  a[(nw-1),] <- 0.5

  rextranot<-rextra
  rextranot[(nwextra-nw+1):nrow(rextranot),(nwextra-nw+1):nrow(rextranot)]<-0

  sextra <- calculate_s(ncol=nwextra-nw, nrow=nwextra, qextra, rextranot, R, T)
  dextra <- sextra[((nwextra-nw+1):nrow(sextra)),1]
  dextra <- dextra[1]-dextra
  dextra <- dextra[2:nw]


  dmatrix <- data.frame(matrix(ncol=1, nrow=(nw-1)))
  dmatrix[,] <-0
  dmatrix[1:5,1] <-dextra[1:5]-dextra[6]
  dmatrix[(nw-1),1]<-qtot

  QQ<-data.frame(matrix(ncol=1, nrow=length(lines)))

  main <- function(qn) {
    resultnew<-rep(NA,length(qn))
    q<-data.frame(matrix(ncol=1, nrow=(nw-1)))
    q[1:(nw-1),1] <- qn[1:(nw-1)]
    QQsquare <- calculate_QQsquare(Cmatrix, q);
    result<-2 * as.matrix(a) %*% as.matrix(q) + 3 * as.matrix(ct) %*% as.matrix(kx) %*% QQsquare -as.matrix(dmatrix)
    resultnew[1:(nw-1)]<-result[1:(nw-1),1]
    return<- resultnew
  }

  newans<- nleqslv::nleqslv(xstart, main, control=list(trace=1,btol=.01,delta="newton"))
  qresult<-newans[["x"]]

  QQsquare<- calculate_QQsquare(Cmatrix, qresult)
  hf <- as.matrix(kx) %*% QQsquare

  s <- calculate_s(ncol=nw-1, nrow=nw-1, qresult, rnot, R, T)

  sfinal <- s[1:nrow(s),1] + sextra[(nwextra-nw+2):nwextra,1]

  return<-list(qresult=qresult,hf=hf,s=s,printmap=printmap,
               sfinal=sfinal,kx=as.matrix(kx),QQsquare=QQsquare)
}

calculate_distances <- function(xc, yc, projected, r0, nw) {
  xy<-data.frame(matrix(ncol=2, nrow=nw)) ; xy[[1]] <- xc ; xy[[2]] <- yc
  coords <- cbind(Easting=xy[1], Northing=xy[2])
  test = sp::SpatialPointsDataFrame(coords, xy, proj4string = sp::CRS(projected))
  test@data[[3]] <- 0:(nw-1)

  r<-rgeos::gDistance(test, byid=TRUE)

  i=0; while (i<nw) {i=i+1; r[[i,i]]=r0}

  list(r=r, coords=coords, test=test)
}

calculate_s <- function(ncol, nrow, q, rnot, R, T) {
  sbasic <- data.frame(matrix(ncol=ncol,nrow=nrow))
  s <- data.frame(matrix(ncol=1,nrow=nrow))
  j=0
  while (j<nrow){
    j=j+1 ;   k=0
    while (k<ncol){
      k=k+1
      sbasic[[j,k]]<- (-1/(2*T*pi)) * q[k] * log(rnot[k,j]/R)
    }
    s[[j,1]] <- sum(sbasic[j,])
  }

  s
}

calculate_QQsquare <- function(Cmatrix, q) {
  QQ<- as.matrix(Cmatrix) %*% as.matrix(q)
  QQsquare<- QQ^2
}

render_diagram <- function(test, lines, qextra, nw, testextra) {
  final_map<- test
  final_map@data[[4]] <- 10
  final_map@data[[1,4]] <- 0
  colnames(final_map@data)<-c("x","y","number","points")
  Mypal <- c('#FF0000','#313695')


  fplot <- function(nw) {
    printm<- tmap::tm_shape(final_map) + tmap::tm_symbols(size= 1, col="number", breaks=c(-5 , 0.5, 100),
                                                          style = "fixed",  palette=Mypal, contrast=1, labels = c("tank","wells"))
    i=0
    while(i<(nw-1)){
      i=i+1
      printm<-printm+tmap::qtm(lines[[i]])
    }
    return<-printm
  }


  if(!(sum(qextra)==0)) {
    printmap<-fplot(nw)+tmap::qtm(testextra)
  } else {
    printmap<-fplot(nw)
  }
}


man/examples/example.R
library(wellsolve)

projected <- "+proj=utm +zone=34"
x0 = 650000
y0 = 4400000

# the first value is the tank and the rest are wells
xc <- c() ; xc <- c(0,100,180,100,700,800,900) ; xc <- xc + x0 -10000
yc <- c() ; yc <- c(0,0,0,80,0,0,900) ; yc <- yc + y0 -10000

# xstart is a wild guess about the result.
# This helps the algorithm search for the result in the correct area
xstart<- c(0.250,0.1,0.25,0.1,0.1,0.01)

nw=7 #number of wells + 1 (the tank)
links<- data.frame(matrix(ncol=2, nrow=(nw-1)))
links[1:(nw-1),1]<-1:(nw-1)
links[1,2]<-0 ; links[2,2]<-1 ; links[3,2]<-0
links[4,2]<-2 ; links[5,2]<-4 ; links[6,2]<-2

Dlist<-list()
Dlist[1:6]<-0.05


# number of extra wells
nwextra <- 2
xextra<- c() ; xextra <- c(500,800) ; xextra <- xextra + x0 -10000
yextra<- c() ; yextra <- c(500,300) ; yextra <- yextra + y0 -10000
qextra<- c() ; qextra[1]<- 0; qextra[2]<- 0

result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)

# Case A
Dlist[1:6]<-2
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)

# Case LC1
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)

# Case B
Dlist[1:6]<-0.4
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)

# Case C
qextra[1] <- 0.1
qextra[2] <- 0.4
result<-wellsolve(projected, T=0.0025 ,R=3000, ff=0.03, qtot<-0.5, r0=0.200,
                  Dlist,qextra,nw,xc,yc,links,nwextra,xextra,yextra, xstart)
